feat: add `math/base/special/modff`

anandkaranubc opened the pull request:
Progresses https://github.com/stdlib-js/stdlib/issues/649

## Description

> What is the purpose of this pull request?

This pull request:

-  Adds the complete implementation for `math/base/special/modff`

## Related Issues

> Does this pull request have any related issues?

This pull request:

-   Progresses https://github.com/stdlib-js/stdlib/issues/649

## Questions

> Any questions for reviewers of this pull request?

No.

## Other

> Any other information relevant to this pull request? This may include screenshots, references, and/or implementation notes.

No.

## Checklist

> Please ensure the following tasks are completed before submitting this pull request.

-   [x] Read, understood, and followed the [contributing guidelines][contributing].

* * *

@stdlib-js/reviewers

[contributing]: https://github.com/stdlib-js/stdlib/blob/develop/CONTRIBUTING.md

anandkaranubc committed:
feat: add `math/base/special/modff`

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: passed
  - task: lint_package_json
    status: passed
  - task: lint_repl_help
    status: passed
  - task: lint_javascript_src
    status: passed
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: passed
  - task: lint_javascript_tests
    status: passed
  - task: lint_javascript_benchmarks
    status: passed
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: passed
  - task: lint_c_examples
    status: passed
  - task: lint_c_benchmarks
    status: passed
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: passed
  - task: lint_typescript_tests
    status: passed
  - task: lint_license_headers
    status: passed
---

anandkaranubc commented:
FreeBSD implementation has similar logic.

https://github.com/freebsd/freebsd-src/blob/main/lib/msun/src/s_modff.c

anandkaranubc committed:
Update benchmark.native.js

Signed-off-by: Karan Anand <119553199+anandkaranubc@users.noreply.github.com>

anandkaranubc committed:
Merge branch 'develop' into feat/modff

anandkaranubc committed:
bench: use `random_uniform`

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: na
  - task: lint_package_json
    status: na
  - task: lint_repl_help
    status: na
  - task: lint_javascript_src
    status: na
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: na
  - task: lint_javascript_tests
    status: na
  - task: lint_javascript_benchmarks
    status: na
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: na
  - task: lint_c_examples
    status: na
  - task: lint_c_benchmarks
    status: passed
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: na
  - task: lint_typescript_tests
    status: na
  - task: lint_license_headers
    status: passed
---

Planeshifter commented on lib/node_modules/@stdlib/math/base/special/modff/manifest.json:
```diff
@@ -0,0 +1,96 @@
+    {
+      "task": "benchmark",
+      "src": [
+        "./src/main.c"
+      ],
+      "include": [
+        "./include"
+      ],
+      "libraries": [],
+      "libpath": [],
+      "dependencies": [
+        "@stdlib/math/base/assert/is-nanf",
+        "@stdlib/number/float32/base/from-word",
+        "@stdlib/number/float32/base/to-word",
+        "@stdlib/constants/float32/pinf",
+        "@stdlib/constants/float32/exponent-bias",
+        "@stdlib/constants/float32/exponent-mask",
+        "@stdlib/constants/float32/significand-mask",
+        "@stdlib/number/float32/base/to-word"
+      ]
+    },
+    {
+      "task": "examples",
+      "src": [
+        "./src/main.c"
+      ],
+      "include": [
+        "./include"
+      ],
+      "libraries": [],
+      "libpath": [],
+      "dependencies": [
+        "@stdlib/math/base/assert/is-nanf",
+        "@stdlib/number/float32/base/from-word",
+        "@stdlib/number/float32/base/to-word",
+        "@stdlib/constants/float32/pinf",
+        "@stdlib/constants/float32/exponent-bias",
+        "@stdlib/constants/float32/exponent-mask",
+        "@stdlib/constants/float32/significand-mask",
+        "@stdlib/number/float32/base/to-word"
```
```suggestion
        "@stdlib/napi/argv",
        "@stdlib/napi/argv-float",
        "@stdlib/napi/argv-float32array",
        "@stdlib/napi/export",
        "@stdlib/math/base/assert/is-nanf",
        "@stdlib/number/float32/base/from-word",
        "@stdlib/number/float32/base/to-word",
        "@stdlib/constants/float32/pinf",
        "@stdlib/constants/float32/exponent-bias",
        "@stdlib/constants/float32/exponent-mask",
        "@stdlib/constants/float32/significand-mask"
      ]
    },
    {
      "task": "benchmark",
      "src": [
        "./src/main.c"
      ],
      "include": [
        "./include"
      ],
      "libraries": [],
      "libpath": [],
      "dependencies": [
        "@stdlib/math/base/assert/is-nanf",
        "@stdlib/number/float32/base/from-word",
        "@stdlib/number/float32/base/to-word",
        "@stdlib/constants/float32/pinf",
        "@stdlib/constants/float32/exponent-bias",
        "@stdlib/constants/float32/exponent-mask",
        "@stdlib/constants/float32/significand-mask"
      ]
    },
    {
      "task": "examples",
      "src": [
        "./src/main.c"
      ],
      "include": [
        "./include"
      ],
      "libraries": [],
      "libpath": [],
      "dependencies": [
        "@stdlib/math/base/assert/is-nanf",
        "@stdlib/number/float32/base/from-word",
        "@stdlib/number/float32/base/to-word",
        "@stdlib/constants/float32/pinf",
        "@stdlib/constants/float32/exponent-bias",
        "@stdlib/constants/float32/exponent-mask",
        "@stdlib/constants/float32/significand-mask"
```

Planeshifter replied:
```suggestion
        "@stdlib/constants/float32/significand-mask"
```

Planeshifter commented on lib/node_modules/@stdlib/math/base/special/modff/manifest.json:
```diff
@@ -0,0 +1,96 @@
+        "@stdlib/constants/float32/exponent-bias",
+        "@stdlib/constants/float32/exponent-mask",
+        "@stdlib/constants/float32/significand-mask",
+        "@stdlib/number/float32/base/to-word"
```
Duplicated entry:

```suggestion
        "@stdlib/napi/argv",
        "@stdlib/napi/argv-float",
        "@stdlib/napi/argv-float32array",
        "@stdlib/napi/export",
        "@stdlib/math/base/assert/is-nanf",
        "@stdlib/number/float32/base/from-word",
        "@stdlib/number/float32/base/to-word",
        "@stdlib/constants/float32/pinf",
        "@stdlib/constants/float32/exponent-bias",
        "@stdlib/constants/float32/exponent-mask",
        "@stdlib/constants/float32/significand-mask"
```

Planeshifter commented on lib/node_modules/@stdlib/math/base/special/modff/manifest.json:
```diff
@@ -0,0 +1,96 @@
+        "@stdlib/constants/float32/exponent-bias",
+        "@stdlib/constants/float32/exponent-mask",
+        "@stdlib/constants/float32/significand-mask",
+        "@stdlib/number/float32/base/to-word"
```
```suggestion
        "@stdlib/math/base/assert/is-nanf",
        "@stdlib/number/float32/base/from-word",
        "@stdlib/number/float32/base/to-word",
        "@stdlib/constants/float32/pinf",
        "@stdlib/constants/float32/exponent-bias",
        "@stdlib/constants/float32/exponent-mask",
        "@stdlib/constants/float32/significand-mask"
```

Planeshifter commented on lib/node_modules/@stdlib/math/base/special/modff/src/main.c:
```diff
@@ -0,0 +1,111 @@
+*
+* @param x           input value
+* @param frac        destination to store the normalized fraction
+* @param integral    destination to store the exponent
```
```suggestion
* @param integral    destination pointer for the integral part
* @param frac        destination pointer for the fractional part
```

Planeshifter commented on lib/node_modules/@stdlib/math/base/special/modff/src/main.c:
```diff
@@ -0,0 +1,111 @@
+		word &= ( ~i );
+
+		// Generate the integral part:
+		j = (float)i;
```
Isn't `j` assigned a value in L99, so why is this needed here?

Ref: https://github.com/stdlib-js/stdlib/tree/develop/lib/node_modules/%40stdlib/number/float32/base/from-word#stdlib_base_float32_from_word-word-x-

Planeshifter commented on lib/node_modules/@stdlib/math/base/special/modff/README.md:
```diff
@@ -0,0 +1,215 @@
+The function accepts the following arguments:
+
+-   **x**: `[in] float` input value.
+-   **frac**: `[out] float*` destination for the integral part.
+-   **exp**: `[out] float*` destination for the fractional part.
```
```suggestion
-   **integral**: `[out] float*` destination for the integral part.
-   **frac**: `[out] float*` destination for the fractional part.
```

Planeshifter commented on lib/node_modules/@stdlib/math/base/special/modff/docs/types/index.d.ts:
```diff
@@ -0,0 +1,79 @@
+	* var parts = modff( 3.14 );
+	* // returns [ 3.0, 0.1400001049041748 ]
+	*/
+	( x: number ): Array<number>;
```
Is there any reason to not type the return value as a two-element tuple here?

```suggestion
	( x: number ): [ number, number ];
```

cc @kgryte

kgryte replied:
@Planeshifter No. You originally typed as a `Collection` (ref: https://github.com/stdlib-js/stdlib/commit/6921c387952e3035aebe0ec6e6b223758c4b5371). That was carried over in a subsequent refactoring.

Planeshifter reviewed (changes requested):
In great shape overall, but noticed a few small things we should address.

anandkaranubc committed:
chore: remove duplicated entries from `manifest.json`

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: na
  - task: lint_package_json
    status: na
  - task: lint_repl_help
    status: na
  - task: lint_javascript_src
    status: na
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: na
  - task: lint_javascript_tests
    status: na
  - task: lint_javascript_benchmarks
    status: na
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: na
  - task: lint_c_examples
    status: na
  - task: lint_c_benchmarks
    status: na
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: na
  - task: lint_typescript_tests
    status: na
  - task: lint_license_headers
    status: passed
---

anandkaranubc committed:
Merge branch 'develop' into feat/modff

anandkaranubc committed:
refactor: use tuple for return type

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: na
  - task: lint_package_json
    status: na
  - task: lint_repl_help
    status: na
  - task: lint_javascript_src
    status: na
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: na
  - task: lint_javascript_tests
    status: na
  - task: lint_javascript_benchmarks
    status: na
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: na
  - task: lint_c_examples
    status: na
  - task: lint_c_benchmarks
    status: na
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: passed
  - task: lint_typescript_tests
    status: passed
  - task: lint_license_headers
    status: passed
---

anandkaranubc committed:
chore: apply suggestions from review

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: passed
  - task: lint_package_json
    status: na
  - task: lint_repl_help
    status: na
  - task: lint_javascript_src
    status: na
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: na
  - task: lint_javascript_tests
    status: na
  - task: lint_javascript_benchmarks
    status: na
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: passed
  - task: lint_c_examples
    status: na
  - task: lint_c_benchmarks
    status: na
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: na
  - task: lint_typescript_tests
    status: na
  - task: lint_license_headers
    status: passed
---

Planeshifter reviewed (approved):
Thanks, @anandkaranubc!
